torch/_inductor/fx_passes/post_grad.py

```diff
@@ -216,6 +217,54 @@ def is_valid_mm_plus_mm(match: Match):
     return True
 
 
+def scatter_upon_allzero_extra_check(m):
+    allzero_shape = m.kwargs["shape"]
+    selector = m.kwargs["selector"]
+    val = m.kwargs["val"]
+
+    # Only match scatter against 2D tensors for now.
+    if len(allzero_shape) != 2:
+        return False
+    if not isinstance(val, float):
+        return False
+
+    selector_ft = selector.meta["val"]
+    if selector_ft.dim() == 2 and selector_ft.size(1) != 1:
+        return False
+    return True
+
+
+@register_lowering_pattern(
+    CallFunction(
+        aten.scatter,
+        CallFunction(aten.full, KeywordArg("shape"), 0.0, dtype=KeywordArg("dtype")),
+        1,  # dim
+        KeywordArg("selector"),
+        KeywordArg("val"),  # scalar value
+    ),
+    extra_check=scatter_upon_allzero_extra_check,
+)
+def scatter_upon_allzero(match: Match, shape, dtype, selector, val):
+    from torch._inductor import lowering
+
+    if len(selector.get_size()) == 2:
```

Comment:
lowering pattern only works in post-grad pass, right?

I'm worried that if we express the replacement in aten, the default lowering rule of scatter would make it not fusable to consumer nodes.

Reply:
```
def decomp(selector, val):
    x_dim = torch.arange(selector.shape[0])
    y_dim = torch.arange(selector.shape[1])
    return torch.where(selector[x_dim] == y_dim, val, 0)
```

does something along those work ? need to think abt it

Reply:
nope, selector is a 1D tensor here.

Reply:
I still think we can do this. For example:


```
import torch
torch.set_default_device("cuda")

# Example usage
M, N = 1024, 2048
x = torch.randint(0, N, (M,), dtype=torch.int64)

def f(x):
    y = torch.full([M, N], 3.14, dtype=torch.float)
    y.scatter_(1, x.unsqueeze(1), 2.718)
    return y

def pointwise_scatter(x):
    y = torch.full((M, N), 3.14, dtype=torch.float, device=x.device)
    range_tensor = torch.arange(N, device=x.device).view(1, -1)
    x_expanded = x.unsqueeze(1).expand(-1, N)
    condition = (x_expanded == range_tensor)
    return torch.where(condition, torch.full_like(y, 2.718), y)
    
torch.testing.assert_allclose(optimized_scatter(x), f(x))
```

Reply:
Nice to see this can be expressed as aten ops! I didn't realize that we can use `arange` like this to generate all indices .

But as I understand, moving the pattern to joint graph is generally good since the partitioner may figure out something smarter. But it's may also be just neutral in practice. Since we've already done thorough tests on this PR, should we get this version in and do another round of tests with the equivalent graph pattern you proposed in a followup PR?

Reply:
Sure, we can do as a follow up (maybe a `good first issue` / `internal ramp up task`)

Reply:
Here it is: https://github.com/pytorch/pytorch/issues/129449